llvm/include/llvm-c/ExternC.h

```diff
@@ -36,4 +36,18 @@
 #define LLVM_C_EXTERN_C_END LLVM_C_STRICT_PROTOTYPES_END
 #endif
 
+#ifndef LLVM_HAS_ATTRIBUTE
+#if defined(__has_attribute)
+#define LLVM_HAS_ATTRIBUTE(x) __has_attribute(x)
+#else
+#define LLVM_HAS_ATTRIBUTE(x) 0
+#endif
+#endif
+
+#if LLVM_HAS_ATTRIBUTE(warn_unused_result) && defined(__clang__)
```

Comment:
In the Phab review, you commented:

> I guess this also needs a more narrow macro check to verify the clang version (since it seems we support back to clang 5, and this only works on clang 15 and above) - I can't seem to find an example of a macro check for clang version in any of our Compiler.h and similar macro-y/compat headers in LLVM - do you happen to know of an example off-hand?

Do we still need to do this? We don't have a macro for version checking Clang in Compiler.h, but we probably could use one. We have `LLVM_GNUC_PREREQ` and `LLVM_MSC_PREREQ`, so adding `LLVM_CLANG_PREREQ` would be pretty natural.

Reply:
> Do we still need to do this? 

I think the original rationale still applies - [we support building LLVM with Clang 5.0](https://llvm.org/docs/GettingStarted.html#host-c-toolchain-both-compiler-and-standard-library) and support for `__attribute__((warn_unused_result))` was added in Clang 15. Prior to that clang warns that the attribute is used out of place - so we'd want to conditionalize the attribute so as not to warn/werror-break the build when using a supported (>= 5) clang < 15.

>  We don't have a macro for version checking Clang in Compiler.h, but we probably could use one. We have LLVM_GNUC_PREREQ and LLVM_MSC_PREREQ, so adding LLVM_CLANG_PREREQ would be pretty natural.

Fair - though this one can't go in Compiler.h because it needs to be down in llvm-c and I don't think Compiler.h is that low-level. So we might have to pull the other _PREREQ macros down into llvm-c for symmetry.